rts/motoko-rts/src/persistence.rs

```diff
@@ -0,0 +1,223 @@
+        use crate::gc::incremental::IncrementalGC;
+        debug_assert!(!self.is_initialized());
+        (*self).fingerprint = FINGERPRINT;
+        (*self).version = VERSION;
+        (*self).stable_actor = DEFAULT_VALUE;
+        (*self).stable_type = TypeDescriptor::default();
+        (*self).incremental_gc_state = IncrementalGC::initial_gc_state(mem, HEAP_START);
+        (*self).upgrade_instructions = 0;
+    }
+}
+
+/// Initialize fresh persistent memory after the canister installation or
+/// reuse the persistent memory on a canister upgrade.
+#[cfg(feature = "ic")]
+pub unsafe fn initialize_memory<M: Memory>(mem: &mut M) {
+    mem.grow_memory(HEAP_START as u64);
+    let metadata = PersistentMetadata::get();
+    if metadata.is_initialized() {
+        metadata.check_version();
+    } else {
+        metadata.initialize(mem);
+    }
+}
+
+/// Returns the stable sub-record of the actor of the upgraded canister version.
+/// Returns scalar 0 if no actor is stored after on a fresh memory.
+#[no_mangle]
+pub unsafe extern "C" fn load_stable_actor() -> Value {
+    let metadata = PersistentMetadata::get();
+    (*metadata).stable_type.assert_initialized();
+    (*metadata).stable_actor.forward_if_possible()
+}
+
+/// Save the stable sub-record, the stable variables, of a canister before an upgrade.
+#[ic_mem_fn]
+pub unsafe fn save_stable_actor<M: Memory>(mem: &mut M, actor: Value) {
+    assert!(actor != DEFAULT_VALUE);
+    let metadata: *mut PersistentMetadata = PersistentMetadata::get();
+    (*metadata).stable_type.assert_initialized();
+    let location = &mut (*metadata).stable_actor as *mut Value;
+    write_with_barrier(mem, location, actor);
+}
+
+/// Free the stable actor sub-record after a completed upgrade.
+/// Allow garbage collection for unused stable variables that are no longer declared
+/// in the new program version.
+#[ic_mem_fn]
+pub unsafe fn free_stable_actor<M: Memory>(mem: &mut M) {
+    let metadata: *mut PersistentMetadata = PersistentMetadata::get();
+    let location = &mut (*metadata).stable_actor as *mut Value;
+    write_with_barrier(mem, location, DEFAULT_VALUE);
+}
+
+#[cfg(feature = "ic")]
+/// GC root pointer required for GC marking and updating.
+pub(crate) unsafe fn stable_actor_location() -> *mut Value {
+    let metadata = PersistentMetadata::get();
+    &mut (*metadata).stable_actor as *mut Value
+}
+
+/// Determine whether an object contains a specific field.
+/// Used for upgrading to an actor with additional stable fields.
+#[no_mangle]
+#[cfg(feature = "ic")]
+pub unsafe extern "C" fn contains_field(actor: Value, field_hash: u32) -> bool {
+    use crate::constants::WORD_SIZE;
+
+    let object = actor.as_object();
+    let hash_blob = (*object).hash_blob.as_blob();
+    assert_eq!(hash_blob.len().as_u32() % WORD_SIZE, 0);
+    let number_of_fields = hash_blob.len().as_u32() / WORD_SIZE;
+    let mut current_address = hash_blob.payload_const() as u32;
```

Comment:
```suggestion
    let mut current_address = hash_blob.payload_const() as usize;
```
Perhaps?

Reply:
Indeed, I changed this in the 64-bit port to `usize` (the next higher PR). Here, it is just following the local pattern, where `WORD_SIZE` is `u32` and then logic of pointer advancing is then simpler.